File: concepts/data structures/implementation/tree/bt_traversal/depth_first/python/bt_post_order.py
from bt_base.python.binary_tree import BinaryNode


def walk(curr, path):
    if curr is None:
        return path

    # Recursion
    # Pre
    # Recurse
    walk(curr.left, path)
    walk(curr.right, path)

    # Post
    path.append(curr.value)
    return path


def post_order_search(head):
    return walk(head, [])


def test_post_order_search():
    tree = BinaryNode(
        1,
        BinaryNode(2, BinaryNode(4), BinaryNode(5)),
        BinaryNode(3),
    )
    result = post_order_search(tree)

    print(f"Post-order: {result} | Expected: [4, 5, 2, 3, 1]")


test_post_order_search()

File: concepts/data structures/implementation/tree/bt_traversal/depth_first/typescript/bt_post_order.ts
import { BinaryNode } from "../../../bt_base/typescript/binary_tree";

function walk(curr: BinaryNode<number> | null, path: number[]): number[] { 
  if (!curr) return path;

  // recursion
  // pre
  // recurse
  walk(curr.left, path);
  walk(curr.right, path);

  // post
  path.push(curr.value);
  return path;
}

function post_order_search(head: BinaryNode<number>): number[] {
  return walk(head, []);
}

function testPostOrderSearch(): void {
  const tree = new BinaryNode(
    1,
    new BinaryNode(2, new BinaryNode(4), new BinaryNode(5)),
    new BinaryNode(3),
  );
  const result = post_order_search(tree);

  console.log(`Post-order: ${result.join(", ")} | Expected: 4, 5, 2, 3, 1`);
}

testPostOrderSearch();
